lib/inventory/metadata.go

```diff
@@ -0,0 +1,366 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inventory
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strconv"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// Service names for UpstreamInventoryAgentMetadata
+const (
+	nodeService           = "node"
+	kubeService           = "kube"
+	appService            = "app"
+	dbService             = "db"
+	windowsDesktopService = "windows_desktop"
+)
+
+// This regexp is used to validate if the host architecture fetched has the
+// expected format.
+var matchHostArchitecture = regexp.MustCompile(`^\w+$`)
+
+// fetchConfig contains the configuration used by the fetchAgentMetadata method.
+type fetchConfig struct {
+	ctx context.Context
+	// hello is the initial upstream hello message.
+	hello proto.UpstreamInventoryHello
+	// getenv is the method called to retrieve an environment
+	// variable.
+	// It is configurable so that it can be mocked in tests.
+	getenv func(name string) string
+	// readFile is the method called to read a file.
+	// It is configurable so that it can be mocked in tests.
+	readFile func(name string) ([]byte, error)
+	// execCommand is the method called to execute a command.
+	// It is configurable so that it can be mocked in tests.
+	execCommand func(name string, args ...string) ([]byte, error)
+	// httpDo is the method called to perform an http request.
+	// It is configurable so that it can be mocked in tests.
+	httpDo func(req *http.Request) (*http.Response, error)
+	// kubeClient is a kubernetes client used to retrieve the
+	// server version.
+	// It is configurable so that it can be mocked in tests.
+	kubeClient kubernetes.Interface
+}
+
+// setDefaults sets the values of readFile and execCommand to the ones in the
+// standard library. Having these two methods configurable allows us to mock
+// them in tests.
+func (c *fetchConfig) setDefaults() {
+	if c.ctx == nil {
+		c.ctx = context.Background()
+	}
+	if c.getenv == nil {
+		c.getenv = os.Getenv
+	}
+	if c.readFile == nil {
+		c.readFile = os.ReadFile
+	}
+	if c.execCommand == nil {
+		c.execCommand = func(name string, args ...string) ([]byte, error) {
+			return exec.Command(name, args...).Output()
+		}
+	}
+	if c.httpDo == nil {
+		c.httpDo = func(req *http.Request) (*http.Response, error) {
+			client := &http.Client{}
```

Comment:
We seem to use this throughout the codebase.
```suggestion
			client := defaults.HTTPClient()
```